bpo-17611. Move unwinding of stack for "pseudo exceptions" from interpreter to compiler.

serhiy-storchaka opened the pull request:
<!-- issue-number: bpo-17611 -->
https://bugs.python.org/issue17611
<!-- /issue-number -->

pitrou committed:
bpo-17611: Move unwinding of stack from interpreter to compiler

pitrou committed:
Cleanup

pitrou committed:
Improve stack size calculation for SETUP_{EXCEPT,FINALLY}

pitrou committed:
Update frozen importlib

pitrou committed:
Remove useless code

pitrou committed:
Don't predict FOR_ITER as it breaks tracing without computed gotos

pitrou committed:
Remove dead code and fix tracing tests

pitrou committed:
Fix test_dis

pitrou committed:
Add JUMP_FINALLY to avoid finally block duplication

pitrou committed:
Remove finally block duplication

pitrou committed:
More precise computation of code object stack size

pitrou committed:
JUMP_FINALLY pushes 6 values to be consistent with the effect of an exception

pitrou committed:
Exact stack size computation by popping stale exception state in RERAISE

pitrou committed:
Remove the now pointless POP_MANY

pitrou committed:
Add comment for RERAISE

pitrou committed:
Remove JUMP_FINALLY

pitrou committed:
Remove last block duplication

pitrou committed:
Get rid of END_ITER

pitrou committed:
Fix comments in frameobject.c

pitrou committed:
Add stack size tests

serhiy-storchaka committed:
Fix test_dis.

serhiy-storchaka committed:
Simplify handling of 'for' loops in frame_setlineno() and fix it for long opargs.

serhiy-storchaka committed:
Do not duplicate the finally block.

serhiy-storchaka committed:
Refactor the frame unwinding code.

serhiy-storchaka committed:
Refactoring and optimizations in the compiler.

serhiy-storchaka committed:
Fix frame_setlineno().

serhiy-storchaka committed:
Rename PUSH_NO_EXCEPT to BEGIN_FINALLY.

serhiy-storchaka committed:
Restore SETUP_WITH and SETUP_ASYNC_WITH.

serhiy-storchaka committed:
Fix stack depth calculation.

serhiy-storchaka committed:
Remove SETUP_EXCEPT.

serhiy-storchaka committed:
Fix rebase error.

serhiy-storchaka committed:
Change codes of new instructions.

serhiy-storchaka committed:
Update the magic number.

serhiy-storchaka committed:
Remove debug print.

serhiy-storchaka committed:
Polishing.

serhiy-storchaka committed:
Document changes.

serhiy-storchaka committed:
Make CALL_FINALLY a jump instruction, but exclude it from stack depth walk.

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Fix jump tests: jumping into a while block is possible now.

pitrou commented on Python/compile.c:
```diff
@@ -81,11 +81,17 @@ It's called a frame block to distinguish it from a basic block in the
+enum fblocktype { WHILE_LOOP, FOR_LOOP, EXCEPT, FINALLY_TRY, FINALLY_END,
+                  WITH, ASYNC_WITH, HANDLER_CLEANUP };
+
+struct fblockinfo;
```
This forward decl doesn't look necessary.

pitrou commented on Python/compile.c:
```diff
@@ -2469,9 +2613,8 @@ compiler_continue(struct compiler *c)
-    Pops a variable number of entries from the *value* stack
-    and re-raises the exception they specify.  The number of
-    entries popped depends on the (pseudo) exception type.
+    Pops 3 entries from the *value* stack and re-raises the exception
```
This doesn't match the END_FINALLY description in `dis.rst` above.

pitrou commented on Lib/test/test_compile.py:
```diff
@@ -710,5 +710,102 @@ def test_func_and(self):
+            """
+        self.check_stack_size(snippet)
+
+    # This one unfortunately "leaks" a few stack slots for each snippet
```
Do you have any idea how to solve this one?

serhiy-storchaka replied:
No, I don't have any idea. But since this test is failed with the current code I don't worry. Maybe bpo-31113 will fix it.

serhiy-storchaka committed:
Polishing.

serhiy-storchaka committed:
Fix unrolling the finally block.

serhiy-storchaka committed:
Optimize 'return' with constant value.

serhiy-storchaka committed:
Do not eliminate END_FINALLY in peepholer.

serhiy-storchaka committed:
Minimize changes in PyCompile_OpcodeStackEffect().

serhiy-storchaka committed:
Document POP_FINALLY and add tests for 'break' and 'return' in 'finally' clause.

serhiy-storchaka committed:
Polishing.

serhiy-storchaka committed:
Minimize changes in WITH_CLEANUP_START and WITH_CLEANUP_START.

pitrou commented on Lib/opcode.py:
```diff
@@ -214,5 +211,7 @@ def jabs_op(name, op):
 name_op('LOAD_METHOD', 160)
 def_op('CALL_METHOD', 161)
+jrel_op('CALL_FINALLY', 162)
+jrel_op('POP_FINALLY', 163)
```
Why is POP_FINALLY a relative jump?

serhiy-storchaka replied:
Good catch! This is an error.

pitrou commented on Python/compile.c:
```diff
@@ -1002,27 +1008,33 @@ PyCompile_OpcodeStackEffect(int opcode, int oparg)
-            return 6; /* can push 3 values for the new exception
-                + 3 others for the previous exception state */
+        case POP_FINALLY:
+            return -6;
```
Why is this statically -6? This would deserve a comment.

serhiy-storchaka replied:
Done.

pitrou commented on Doc/library/dis.rst:
```diff
@@ -670,11 +668,45 @@ iterations of the loop.
 
+.. opcode:: POP_FINALLY (preserve_tos)
+
+   Cleans up the value stack and the block stack.  If *preserve_tos* is not
```
For the sake of static bytecode analyzers, perhaps it would be better to have two separate opcodes: one preserving TOS, the other not.

serhiy-storchaka replied:
Wouldn't this just complicate the code and the documentation?

pitrou replied:
The code wouldn't be much more complicated (you can share the same implementation in the eval loop). OTOH I'm not sure it's necessary.

Other question: why are both POP_FINALLY and END_FINALLY needed? The documentation should be a bit clearer.

serhiy-storchaka replied:
`END_FINALLY` re-raises an exception, `POP_FINALLY 0` just drops it. `END_FINALLY` jumps to the return address, `POP_FINALLY 0` ignores it. And `POP_FINALLY 1` preserves TOS.

I don't know what to add to the documentation. Can you provide your changes?

If push 6 `None` instead of 1 `NULL` in `BEGIN_FINALLY`,  and push an exception block in `BEGIN_FINALLY`, and update other instructions accordingly, and add additional hacks to stack effect calculation, `POP_FINALLY 0` could be replaced with:
```
POP_TOP
POP_TOP
POP_TOP
POP_EXCEPT
```
and `POP_FINALLY 1` could be replaced with:
```
ROT_FOUR
POP_TOP
POP_TOP
POP_TOP
ROT_FOUR
POP_EXCEPT
```
This was my initial solution for @nascheme's examples. But this would add too much unneeded overhead and complication. The single new opcode has lesser cost.

pitrou replied:
> END_FINALLY re-raises an exception, POP_FINALLY 0 just drops it. END_FINALLY jumps to the return address, POP_FINALLY 0 ignores it. 

Yes... So what I don't understand is why both are needed. In which case do you need to drop an exception?

serhiy-storchaka replied:
'break' and 'return' in 'finally' clause.

'continue' is specially prohibited, but it can be allowed too after merging this PR.

pitrou replied:
Thank you. Can you add some explanation of this in the docs?

serhiy-storchaka replied:
Done.

serhiy-storchaka committed:
Address review comments.

serhiy-storchaka committed:
Disable FOR_ITER jump optimization.
It fooled the code for detecting boundaries of blocks in the frame
f_lineno setter.
Remove also jump optimizations which never happend.

serhiy-storchaka committed:
Reorder tests.

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Clarify the use case of POP_FINALLY.

serhiy-storchaka committed:
bpo-24340: Fix estimation of the code stack size.

serhiy-storchaka committed:
Add new tests for stack effect (ported from #5076).

serhiy-storchaka committed:
Add comments.

pitrou commented on Doc/library/dis.rst:
```diff
@@ -960,22 +998,19 @@ All of the following opcodes use their arguments.
 
-   Pushes a try block from a try-except clause onto the block stack. *delta*
-   points to the finally block.
+   .. versionadded:: 3.7
```
We don't seem to be adding `versionadded` for other new opcodes, do we?

serhiy-storchaka replied:
`versionadded` is always added for new opcodes. Do you know cases where it was omitted?

pitrou replied:
See `SETUP_ASYNC_WITH`. Or the 2.7 docs, which almost always omit it.

serhiy-storchaka replied:
I think it should be added for `SETUP_ASYNC_WITH`. It is used with other new opcodes added before and after `SETUP_ASYNC_WITH`.

pitrou commented on Doc/library/dis.rst:
```diff
@@ -670,11 +668,49 @@ iterations of the loop.
+     used to restore the exception state.  An exception handler block is
+     removed from the block stack.
+
+   It is similar to :opcode:`END_FINALLY`, but neither change the bytecode
```
I would say "but doesn't change the bytecode counter nor raise an exception".

serhiy-storchaka replied:
Thanks!

pitrou commented on Python/ceval.c:
```diff
@@ -1869,67 +1865,124 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
+        }
+
+        TARGET(BEGIN_FINALLY) {
+            PUSH(NULL);
```
Explain this is in a comment here?

serhiy-storchaka replied:
Done.

pitrou commented on Python/ceval.c:
```diff
@@ -1869,67 +1865,124 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
+        }
+
+        TARGET(POP_FINALLY) {
+            PyObject *res = NULL;
```
I would expect a similar comment as at the to top of END_FINALLY here.

serhiy-storchaka replied:
Done.

pitrou commented on Python/compile.c:
```diff
@@ -1002,27 +1008,36 @@ PyCompile_OpcodeStackEffect(int opcode, int oparg)
+             * + 3 others for the previous exception state */
+            return 6;
+        case BEGIN_FINALLY:
+            /* The stack has been reserved by SETUP_FINALLY */
```
Still, it pushes a NULL... ?

serhiy-storchaka replied:
Different approach is used now. The meaning of BEGIN_FINALLY still is not trivial, but it is now explained in a comment.

serhiy-storchaka committed:
Simplify the algorithm, add comments and asserts.

serhiy-storchaka committed:
Merge branch 'master' into stack-size

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'stack-size' into unwind_stack

serhiy-storchaka committed:
Fix stack effect computation.

serhiy-storchaka committed:
Update comments and docs in responce to the review.

serhiy-storchaka committed:
Add more tests.

serhiy-storchaka committed:
Merge branch 'stack-size' into unwind_stack_new

serhiy-storchaka committed:
Add more comments.

serhiy-storchaka committed:
Define all stack effects in a single place.

serhiy-storchaka committed:
Make stack_effect() static.

serhiy-storchaka committed:
Add braces for PEP 7.

serhiy-storchaka committed:
Merge branch 'master' into stack-size

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'stack-size' into unwind_stack

serhiy-storchaka committed:
Add braces for PEP 7.

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Simplify WITH_CLEANUP_START. It is never called after CALL_FINALLY.

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Update the magic number.

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

pitrou reviewed (approved):
Ok, I haven't looked very closely this time, but declare this to look ok :-)

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack

serhiy-storchaka committed:
Merge branch 'master' into unwind_stack
